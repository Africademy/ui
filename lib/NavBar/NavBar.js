"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _colors = require("../style/colors");

var _Select = _interopRequireDefault(require("../Select"));

var _BufferLogo = _interopRequireDefault(require("./BufferLogo"));

var _NavBarMenu = _interopRequireDefault(require("./NavBarMenu/NavBarMenu"));

var _NavBarProducts = _interopRequireDefault(require("./NavBarProducts/NavBarProducts"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  display: flex;\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  height: 64px;\n  width: 100vw;\n  background: ", ";\n  display: flex;\n  justify-content: space-between;\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var NavBarStyled = _styledComponents.default.nav(_templateObject(), _colors.grayDarker);

var NavBarLeft = _styledComponents.default.div(_templateObject2());

var NavBarRight = _styledComponents.default.div(_templateObject3());
/**
 * The NavBar is not generally consumed alone, but instead used by the AppShell component. (This page is WIP. Examples coming soon.)
 */


var NavBar = function NavBar(_ref) {
  var user = _ref.user;
  return _react.default.createElement(NavBarStyled, null, _react.default.createElement(NavBarLeft, null, _react.default.createElement(_BufferLogo.default, null), _react.default.createElement(_NavBarProducts.default, null)), _react.default.createElement(NavBarRight, null, user.menuItems && _react.default.createElement(_Select.default, {
    horizontalOffset: "-16px",
    customButton: function customButton(handleClick) {
      return _react.default.createElement(_NavBarMenu.default, {
        user: user,
        onClick: handleClick
      });
    },
    items: user.menuItems
  })));
};

NavBar.propTypes = {
  user: _propTypes.default.shape({
    name: _propTypes.default.string,
    email: _propTypes.default.string.isRequired,
    avatar: _propTypes.default.string,
    menuItems: _propTypes.default.arrayOf(_propTypes.default.shape({
      id: _propTypes.default.string.isRequired,
      title: _propTypes.default.string.isRequired,
      component: _propTypes.default.node,
      hasDivider: _propTypes.default.bool,
      onItemClick: _propTypes.default.func
    }))
  }).isRequired
};
var _default = NavBar;
exports.default = _default;