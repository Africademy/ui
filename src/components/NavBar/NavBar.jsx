import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Info as InfoIcon, ArrowLeft, Person as PersonIcon, Cross } from '../Icon';

import { gray, blueDarker, grayLight, grayLighter, grayDark, orangeLighter } from '../style/colors';
import {
  fontWeightMedium,
  fontFamily
} from '../style/fonts';

import Select from '../Select';

import BufferLogo from './BufferLogo';
import NavBarMenu from './NavBarMenu/NavBarMenu';
import NavBarProducts from './NavBarProducts/NavBarProducts';

export function getProductPath(baseUrl) {
  const [, productPath] = baseUrl.match(/https*:\/\/(.+)\.buffer\.com/);
  return productPath;
}

function getRedirectUrl(baseUrl) {
  const productPath = getProductPath(baseUrl);
  return `https://${productPath}.buffer.com`;
}

export function getLogoutUrl(baseUrl = '') {
  const productPath = getProductPath(baseUrl);
  return `https://login${
    productPath.includes('local') ? '.local' : ''
  }.buffer.com/logout?redirect=${getRedirectUrl(baseUrl)}`;
}

export function getAccountUrl(baseUrl = '', user) {
  return `https://account.buffer.com?redirect=${getRedirectUrl(
    baseUrl
  )}&username=${encodeURI(user.name)}`;
}

export function getStopImpersonationUrl(baseUrl = '') { 
  const productPath = getProductPath(baseUrl);
  return `https://admin${
    productPath.includes('local') ? '-next.local' : ''
  }.buffer.com/stopImpersonation?redirect=${getRedirectUrl(baseUrl)}`;
}

const NavBarStyled = styled.nav`
  background: #fff;
  border-bottom: 1px solid ${gray};
  border-top: ${props => (props.isImpersonation ? `2px solid ${orangeLighter}` : `none`)};
  box-shadow: 0 1px 10px -5px rgba(0,0,0,.15);
  display: flex;
  height: 56px;
  justify-content: space-between;
  position: relative;
  width: 100vw;
`;

const NavBarLeft = styled.div`
  display: flex;
`;
const NavBarRight = styled.div`
  display: flex;
`;

const NavBarHelp = styled.a`
  align-items: center;
  color: #fff;
  color: ${props => (props.active ? blueDarker : grayDark)};
  display: flex;
  font-size: 16px;
  font-family: ${fontFamily};
  font-weight: ${fontWeightMedium};
  height: 100%;
  padding: 0 24px;
  position: relative;
  text-decoration: none;
  z-index: 2;
  &:hover {
    color: ${props => (props.active ? blueDarker : grayDark)};
    background-color: ${grayLighter};
  }
  cursor: pointer;
`;

const NavBarHelpText = styled.span`
  margin-left: 8px;
`;

const NavBarVerticalRule = styled.div`
  background-color: ${grayLight};
  height: 24px;
  margin-left: -1px;
  margin-right: -1px;
  position: relative;
  top: 50%;
  transform: translateY(-50%);
  width: 1px;
  z-index: 1;
  display: ${props => (props.isImpersonation ? `none` : `block`)};
`;

export function appendMenuItem(ignoreMenuItems, menuItem) {
  if (!ignoreMenuItems) {
    return menuItem;
  }

  return ignoreMenuItems.includes(menuItem.id) ? null : menuItem;
}

/**
 * The NavBar is not consumed alone, but instead is used by the AppShell component. Go check out the AppShell component to learn more.
 */
class NavBar extends React.Component {
  shouldComponentUpdate(nextProps) {
    return nextProps.user.name !== this.props.user.name ||
      nextProps.user.email !== this.props.user.email;
  }

  render() {
    const { activeProduct, user, helpMenuItems, onLogout, isImpersonation } = this.props;
    return (
      <NavBarStyled isImpersonation={isImpersonation}>
        <NavBarLeft>
          <BufferLogo />
          <NavBarVerticalRule />
          <NavBarProducts activeProduct={activeProduct} />
        </NavBarLeft>
        <NavBarRight>
          {helpMenuItems && (
            <Select
              hideSearch
              capitalizeItemLabel={false}
              customButton={handleClick => (
                <NavBarHelp onClick={handleClick}>
                  <InfoIcon />
                  <NavBarHelpText>Help</NavBarHelpText>
                </NavBarHelp>
              )}
              items={helpMenuItems}
              horizontalOffset="-16px"
              xPosition="right"
            />
          )}
          <NavBarVerticalRule isImpersonation={isImpersonation} />
          <Select
            hideSearch
            capitalizeItemLabel={false}
            xPosition="right"
            isImpersonation={isImpersonation}
            customButton={handleClick => (
              <NavBarMenu user={user} onClick={handleClick} isImpersonation={isImpersonation} />
            )}
            items={[
              appendMenuItem(user.ignoreMenuItems, {
                id: 'account',
                title: 'Account',
                icon: <PersonIcon color={gray} />,
                onItemClick: () => {
                  window.location.assign(
                    getAccountUrl(window.location.href, this.props.user)
                  );
                },
              }),
              ...user.menuItems,
              appendMenuItem(user.ignoreMenuItems, isImpersonation ? {
                id: 'Stop Impersonation',
                title: 'Stop Impersonation',
                icon: <Cross color={gray} />,
                hasDivider: user.menuItems && user.menuItems.length > 0,
                onItemClick: () => {
                  window.location.assign(
                    getStopImpersonationUrl(window.location.href)
                  );
                },
              } : {
                id: 'logout',
                title: 'Logout',
                icon: <ArrowLeft color={gray} />,
                hasDivider: user.menuItems && user.menuItems.length > 0,
                onItemClick: () => {
                  if (typeof onLogout === 'function') onLogout();
                  window.location.assign(getLogoutUrl(window.location.href));
                },
              }),
            ].filter(e => e)}
            horizontalOffset="-16px"
          />
        </NavBarRight>
      </NavBarStyled>
    );
  }
}

NavBar.propTypes = {
  /** The currently active (highlighted) product in the `NavBar`, one of `'publish', 'reply', 'analyze'` */
  activeProduct: PropTypes.oneOf(['publish', 'reply', 'analyze']),

  user: PropTypes.shape({
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    /** If missing we will use Gravatar to get the user avatar by email */
    avatar: PropTypes.string,
    /** If missing we will use Gravatar to get the user avatar by email */
    ignoreMenuItems: PropTypes.arrayOf(PropTypes.string),
    menuItems: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired,
        component: PropTypes.func,
        hasDivider: PropTypes.bool,
        onItemClick: PropTypes.func,
      })
    ).isRequired,
  }).isRequired,
  helpMenuItems: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      component: PropTypes.node,
      hasDivider: PropTypes.bool,
      onItemClick: PropTypes.func,
    })
  ),
  isImpersonation: PropTypes.bool,

  onLogout: PropTypes.func
};

NavBar.defaultProps = {
  activeProduct: undefined,
  helpMenuItems: null,
  isImpersonation: false,
  onLogout: undefined
};

export default NavBar;
