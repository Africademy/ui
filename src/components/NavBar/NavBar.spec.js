import snap from 'jest-auto-snapshots';
import NavBar, { getLogoutUrl, appendMenuItem } from './NavBar';

describe('Logout url', () => {
  it('return logout url', () => {
    const baseUrl = 'https://analyze.buffer.com/facebook/overview/4e88a092512f7e1556000000';
    const productPath = 'analyze'
    expect(getLogoutUrl(baseUrl)).toBe(`https://login${productPath.includes('local') ? '.local' : ''}.buffer.com/logout?redirect=https://${productPath}.buffer.com`);
  });

  it('return local logout url', () => {
    const baseUrl = 'https://analyze.local.buffer.com/facebook/overview/4e88a092512f7e1556000000';
    const productPath = 'analyze.local'
    expect(getLogoutUrl(baseUrl)).toBe(`https://login${productPath.includes('local') ? '.local' : ''}.buffer.com/logout?redirect=https://${productPath}.buffer.com`);
  });
});

describe('Append menu items', () => {
  it('should return the item if no match', () => {
    const item = { id: 'foo' };
    expect(appendMenuItem([], item)).toBe(item);

    expect(appendMenuItem(null, item)).toBe(item);
  });

  it('should return null if ignoreMenuItems matches', () => {
    const item = { id: 'foo' };
    expect(appendMenuItem(['foo'], item)).toBe(null);
  });

});


snap(NavBar, './NavBar.jsx');
