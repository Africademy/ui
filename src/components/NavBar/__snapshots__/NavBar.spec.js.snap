// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-auto-snapshots > NavBar Matches snapshot when passed only required props 1`] = `
<styled.nav>
  <styled.div>
    <BufferLogo />
    <NavBarProduct
      activeProduct="publish"
    />
  </styled.div>
  <styled.div>
    <Select
      customButton={[Function]}
      hasCustomAction={false}
      hasIconOnly={false}
      hideSearch={false}
      horizontalOffset="-16px"
      isOpen={false}
      isSplit={false}
      items={
        Array [
          Object {
            "component": <NodeFixture />,
            "hasDivider": true,
            "id": "jest-auto-snapshots String Fixture",
            "onItemClick": [MockFunction],
            "title": "jest-auto-snapshots String Fixture",
          },
        ]
      }
      label=""
      searchPlaceholder="Search"
      shortcutsEnabled={true}
      size="medium"
      type="secondary"
      xPosition="left"
      yPosition="bottom"
    />
  </styled.div>
</styled.nav>
`;
